File: scripts/qa-image-gen.ts
/**
 * QA script for Hermes image generation setup.
 *
 * Diagnoses the current state of image generation capabilities and reports
 * what's missing or misconfigured. Does NOT modify anything — read-only.
 *
 * Auth detection:
 *   1. xAI OAuth (browser auth) → auth.json providers["xai-oauth"]
 *   2. XAI_API_KEY (env / .env file)
 *   3. Neither → guide the user to choose
 *
 * Usage:
 *   npx tsx scripts/qa-image-gen.ts
 *   node --experimental-strip-types scripts/qa-image-gen.ts   (Node >= 23)
 */

import { readFileSync, existsSync } from "fs";
import { join } from "path";
import { homedir } from "os";
import yaml from "js-yaml";

// --- Paths ---

const HERMES_HOME =
  process.env.HERMES_HOME ||
  join(homedir(), "AppData", "Local", "hermes");
const AUTH_FILE = process.env.HERMES_AUTH_FILE || join(HERMES_HOME, "auth.json");
const ENV_FILE = join(HERMES_HOME, ".env");
const CONFIG_FILE = join(HERMES_HOME, "config.yaml");
const HERMES_AGENT_DIR =
  process.env.HERMES_AGENT_DIR ||
  join(homedir(), "AppData", "Local", "hermes", "hermes-agent");

// --- Colors (disabled if not a tty) ---

const isTty = process.stdout.isTTY;
const C = isTty
  ? {
      green: "\x1b[92m",
      yellow: "\x1b[93m",
      red: "\x1b[91m",
      cyan: "\x1b[96m",
      bold: "\x1b[1m",
      reset: "\x1b[0m",
    }
  : { green: "", yellow: "", red: "", cyan: "", bold: "", reset: "" };

function c(color: keyof typeof C, text: string): string {
  return `${C[color]}${text}${C.reset}`;
}

function check(icon: string, msg: string, ok: boolean, detail = ""): void {
  const status = ok ? c("green", "✓") : c("red", "✗");
  let line = `  ${status} ${icon} ${msg}`;
  if (detail) line += `  (${detail})`;
  console.log(line);
}

// --- Auth detection ---

interface AuthInfo {
  method: string;
  provider: string;
  tokenLength?: number;
  keyLength?: number;
  source?: string;
}

function detectOAuth(): AuthInfo | null {
  if (!existsSync(AUTH_FILE)) return null;
  try {
    const data = JSON.parse(readFileSync(AUTH_FILE, "utf-8"));
    const tokens = data?.providers?.["xai-oauth"]?.tokens;
    const at: string = tokens?.access_token || "";
    if (at && at.length > 20) {
      return {
        method: "xAI OAuth (browser auth)",
        provider: "xai-oauth",
        tokenLength: at.length,
      };
    }
  } catch {
    // ignore
  }
  return null;
}

function detectApiKey(): AuthInfo | null {
  let key = "";
  let source = "";

  // Check .env file
  if (existsSync(ENV_FILE)) {
    for (const line of readFileSync(ENV_FILE, "utf-8").split("\n")) {
      const stripped = line.trim();
      if (stripped.startsWith("XAI_API_KEY=")) {
        key = stripped.slice("XAI_API_KEY=".length).trim().replace(/^["']|["']$/g, "");
        source = `.env (${ENV_FILE})`;
        break;
      }
    }
  }

  // Environment variable overrides .env
  const envKey = (process.env.XAI_API_KEY || "").trim();
  if (envKey) {
    key = envKey;
    source = "environment variable";
  }

  if (key && key.length > 5) {
    return {
      method: "API Key",
      provider: "xai",
      keyLength: key.length,
      source,
    };
  }
  return null;
}

// --- Plugin status ---

interface PluginInfo {
  exists: boolean;
  enabled: boolean;
  version: string;
}

function detectPluginStatus(): Record<string, PluginInfo> {
  const pluginNames = ["image_gen/xai", "video_gen/xai"];
  const results: Record<string, PluginInfo> = {};

  // Read enabled list from config.yaml
  let enabledList: string[] = [];
  if (existsSync(CONFIG_FILE)) {
    try {
      const cfg = yaml.load(readFileSync(CONFIG_FILE, "utf-8")) as Record<string, any>;
      enabledList = cfg?.plugins?.enabled ?? [];
    } catch {
      // ignore
    }
  }

  for (const name of pluginNames) {
    const yamlPath = join(HERMES_AGENT_DIR, "plugins", ...name.split("/"), "plugin.yaml");
    const exists = existsSync(yamlPath);
    let version = "?";

    if (exists) {
      try {
        for (const line of readFileSync(yamlPath, "utf-8").split("\n")) {
          if (line.startsWith("version:")) {
            version = line.split(":").slice(1).join(":").trim();
            break;
          }
        }
      } catch {
        // ignore
      }
    }

    results[name] = {
      exists,
      enabled: enabledList.includes(name),
      version,
    };
  }
  return results;
}

// --- Main ---

function main(): void {
  console.log();
  console.log(c("bold", "═══ Hermes Image Generation QA ═══"));
  console.log();

  // 1. Auth detection
  console.log(c("cyan", "▸ Authentication"));
  console.log();

  const oauth = detectOAuth();
  const apiKey = detectApiKey();
  let hasAnyAuth = false;

  if (oauth) {
    check("🔑", `xAI OAuth detected: ${oauth.method}`, true,
      `token present (${oauth.tokenLength} chars)`);
    hasAnyAuth = true;
  } else {
    check("🔑", "xAI OAuth not found", false);
  }

  if (apiKey) {
    check("🗝️", `API Key detected: ${apiKey.method}`, true,
      `key present (${apiKey.keyLength} chars) via ${apiKey.source}`);
    hasAnyAuth = true;
  } else {
    check("🗝️", "XAI_API_KEY not set", false);
  }

  console.log();

  if (!hasAnyAuth) {
    console.log(c("red", "  ⚠ No xAI credentials found. Choose one:"));
    console.log();
    console.log(`    ${c("yellow", "Option A: Browser auth (recommended)")}`);
    console.log("      hermes model    → select xAI Grok OAuth");
    console.log();
    console.log(`    ${c("yellow", "Option B: API key")}`);
    console.log("      Get a key from https://console.x.ai/");
    console.log("      Then: hermes config set XAI_API_KEY <your-key>");
    console.log(`      Or add XAI_API_KEY=... to ${ENV_FILE}`);
    console.log();
    console.log("  Re-run this script after setting up auth.");
    console.log();
  }

  // 2. Plugin status
  console.log(c("cyan", "▸ Plugins"));
  console.log();

  const plugins = detectPluginStatus();
  for (const [name, info] of Object.entries(plugins)) {
    const detail = info.enabled ? "enabled"
      : info.exists ? "installed (not enabled)"
      : "NOT FOUND";
    check("🔌", `${name} (v${info.version})`, info.enabled, detail);
    if (info.exists && !info.enabled) {
      console.log(`      → Enable: hermes plugins enable ${name}`);
    }
  }

  console.log();

  // 3. Image gen provider config
  console.log(c("cyan", "▸ Provider Selection"));
  console.log();

  let activeProvider = "";
  if (existsSync(CONFIG_FILE)) {
    try {
      const cfg = yaml.load(readFileSync(CONFIG_FILE, "utf-8")) as Record<string, any>;
      activeProvider = cfg?.image_gen?.provider ?? "";
    } catch {
      // ignore
    }
  }

  if (activeProvider === "xai") {
    check("⚙️", "image_gen.provider = xai", true, "explicitly configured");
  } else if (activeProvider) {
    check("⚙️", `image_gen.provider = ${activeProvider}`, false,
      `expected "xai" — hermes may use a different backend`);
    console.log(`      → Fix: hermes config set image_gen.provider xai`);
    console.log(c("yellow", "      ⚠ Restart hermes after changing config (new session required)."));
  } else {
    check("⚙️", "image_gen.provider not set", false,
      "falls back to FAL (legacy default)");
    console.log(`      → Fix: hermes config set image_gen.provider xai`);
    console.log(c("yellow", "      ⚠ Restart hermes after changing config (new session required)."));
  }

  console.log();

  // 4. Toolset status
  console.log(c("cyan", "▸ Toolsets"));
  console.log();
  console.log("  Run this to check & enable:");
  console.log("    hermes tools list          # see current status");
  console.log("    hermes tools enable image_gen   # enable image gen");
  console.log("    hermes tools enable video_gen   # enable video gen");
  console.log();

  // 5. Readiness summary
  console.log(c("cyan", "▸ Readiness Summary"));
  console.log();

  const pluginOk = plugins["image_gen/xai"]?.exists ?? false;
  const pluginEnabled = plugins["image_gen/xai"]?.enabled ?? false;

  const providerOk = activeProvider === "xai";

  if (hasAnyAuth && pluginEnabled && providerOk) {
    console.log(c("green", "  ✅ Looks good! Try generating an image:"));
    console.log();
    console.log('    hermes -z "猫の侍を墨絵スタイルで描いて"');
    console.log();
    if (!oauth) {
      console.log(c("yellow", "  💡 Tip: You're using an API key. Consider switching to OAuth"));
      console.log(c("yellow", "     (browser auth) for a smoother experience: hermes model"));
      console.log();
    }
  } else if (hasAnyAuth && pluginEnabled && !providerOk) {
    console.log(c("yellow", "  ⚠ Auth and plugin are ready, but image_gen.provider is not set to \"xai\"."));
    console.log(c("yellow", "    Fix: hermes config set image_gen.provider xai"));
    console.log();
  } else if (hasAnyAuth && !pluginOk) {
    console.log(c("yellow", "  ⚠ Auth is set but the image_gen/xai plugin is not installed."));
    console.log(c("yellow", "    Enable it with: hermes plugins enable image_gen/xai"));
    console.log();
  } else if (hasAnyAuth && !pluginEnabled) {
    console.log(c("yellow", "  ⚠ Plugin is installed but not enabled."));
    console.log(c("yellow", "    Enable it with: hermes plugins enable image_gen/xai"));
    console.log();
  } else if (!hasAnyAuth && pluginOk) {
    console.log(c("yellow", "  ⚠ Plugin is installed but no credentials configured."));
    console.log(c("yellow", "    See auth options above."));
    console.log();
  } else {
    console.log(c("red", "  ✗ Both auth and plugin need setup. See instructions above."));
    console.log();
  }

  // 6. Also check video_gen
  const videoPlugin = plugins["video_gen/xai"];
  if (videoPlugin?.exists && !videoPlugin.enabled) {
    console.log(`  ${c("cyan", "ℹ")} video_gen/xai plugin is also available.`);
    console.log("    Enable: hermes plugins enable video_gen/xai");
    console.log("    Then:   hermes tools enable video_gen");
    console.log();
  } else if (videoPlugin?.enabled) {
    console.log(`  ${c("green", "ℹ")} video_gen/xai is also enabled — you can generate videos too!`);
    console.log();
  }

  console.log(c("bold", "═══════════════════════════════════"));
  console.log();
}

main();

File: scripts/qa-image-gen.py
#!/usr/bin/env python3
"""
QA script for Hermes image generation setup.

Diagnoses the current state of image generation capabilities and reports
what's missing or misconfigured. Does NOT modify anything — read-only.

Auth detection:
  1. xAI OAuth (browser auth) → auth.json providers["xai-oauth"]
  2. XAI_API_KEY (env / .env file)
  3. Neither → guide the user to choose
"""

import json
import os
import sys
from pathlib import Path
from typing import Optional

# --- Paths ---

HERMES_HOME = Path(os.environ.get("HERMES_HOME", Path.home() / ".hermes"))
AUTH_FILE = Path(os.environ.get("HERMES_AUTH_FILE", HERMES_HOME / "auth.json"))
ENV_FILE = HERMES_HOME / ".env"
CONFIG_FILE = HERMES_HOME / "config.yaml"
HERMES_AGENT_DIR = Path(
    os.environ.get("HERMES_AGENT_DIR",
                   Path.home() / "AppData" / "Local" / "hermes" / "hermes-agent")
)

# Colors (disabled if not a tty)
if sys.stdout.isatty():
    C = {
        "green": "\033[92m", "yellow": "\033[93m", "red": "\033[91m",
        "cyan": "\033[96m", "bold": "\033[1m", "reset": "\033[0m",
    }
else:
    C = {k: "" for k in ("green", "yellow", "red", "cyan", "bold", "reset")}


def _c(color: str, text: str) -> str:
    return f"{C[color]}{text}{C['reset']}"


def _check(icon: str, msg: str, ok: bool, detail: str = ""):
    status = _c("green", "✓") if ok else _c("red", "✗")
    line = f"  {status} {icon} {msg}"
    if detail:
        line += f"  ({detail})"
    print(line)


# --- Auth detection ---

def detect_oauth() -> Optional[dict]:
    """Check if xAI OAuth (browser auth) is configured in auth.json."""
    if not AUTH_FILE.exists():
        return None
    try:
        data = json.loads(AUTH_FILE.read_text(encoding="utf-8"))
        tokens = data.get("providers", {}).get("xai-oauth", {}).get("tokens", {})
        at = tokens.get("access_token", "")
        if at and len(at) > 20:
            return {
                "method": "xAI OAuth (browser auth)",
                "provider": "xai-oauth",
                "token_prefix": at[:20] + "...",
                "token_length": len(at),
                "file": str(AUTH_FILE),
            }
    except (json.JSONDecodeError, KeyError, TypeError):
        pass
    return None


def detect_api_key() -> Optional[dict]:
    """Check if XAI_API_KEY is set in .env or environment."""
    key = None
    source = None

    # Check .env file
    if ENV_FILE.exists():
        for line in ENV_FILE.read_text(encoding="utf-8").splitlines():
            stripped = line.strip()
            if stripped.startswith("XAI_API_KEY="):
                key = stripped.split("=", 1)[1].strip().strip('"').strip("'")
                source = f".env ({ENV_FILE})"
                break

    # Check environment variable (overrides .env)
    env_key = os.environ.get("XAI_API_KEY", "").strip()
    if env_key:
        key = env_key
        source = "environment variable"

    if key and len(key) > 5:
        return {
            "method": "API Key",
            "provider": "xai",
            "key_prefix": key[:8] + "...",
            "key_length": len(key),
            "source": source,
        }
    return None


# --- Plugin status ---

def detect_plugin_status() -> dict:
    """Check if image_gen/xai and video_gen/xai plugins exist and are enabled."""
    results = {}

    for plugin_name in ("image_gen/xai", "video_gen/xai"):
        plugin_dir = HERMES_AGENT_DIR / "plugins" / plugin_name.replace("/", os.sep)
        yaml_path = plugin_dir / "plugin.yaml"

        exists = yaml_path.exists()
        enabled = False
        version = "?"

        if exists:
            try:
                content = yaml_path.read_text(encoding="utf-8")
                for line in content.splitlines():
                    if line.startswith("version:"):
                        version = line.split(":", 1)[1].strip()
            except Exception:
                pass

        # Check enabled status from config.yaml
        enabled_path = HERMES_HOME / "config.yaml"
        if enabled_path.exists() and exists:
            try:
                import yaml as _yaml
                _cfg = _yaml.safe_load(enabled_path.read_text(encoding="utf-8"))
                _enabled_list = (_cfg or {}).get("plugins", {}).get("enabled", [])
                enabled = plugin_name in _enabled_list
            except Exception:
                pass

        results[plugin_name] = {
            "exists": exists,
            "enabled": enabled,  # best-effort
            "version": version,
            "yaml": str(yaml_path),
        }

    return results


def detect_toolset_status() -> dict:
    """Check image_gen and video_gen toolset status (enabled/disabled)."""
    results = {}
    config_path = HERMES_HOME / "config.yaml"
    if config_path.exists():
        try:
            content = config_path.read_text(encoding="utf-8")
            # Look for toolset enable/disable patterns
            for ts in ("image_gen", "video_gen"):
                # Default is enabled for image_gen, disabled for video_gen
                results[ts] = {
                    "toolset": ts,
                    # We can't reliably parse YAML here without a lib,
                    # so just note it exists
                }
        except Exception:
            pass
    return results


# --- Main ---

def main():
    print()
    print(_c("bold", "═══ Hermes Image Generation QA ═══"))
    print()

    # 1. Auth detection
    print(_c("cyan", "▸ Authentication"))
    print()

    oauth = detect_oauth()
    api_key = detect_api_key()

    has_any_auth = False

    if oauth:
        _check("🔑", f"xAI OAuth detected: {oauth['method']}", True,
               f"token={oauth['token_prefix']} ({oauth['token_length']} chars)")
        has_any_auth = True
    else:
        _check("🔑", "xAI OAuth not found", False)

    if api_key:
        _check("🗝️", f"API Key detected: {api_key['method']}", True,
               f"prefix={api_key['key_prefix']} via {api_key['source']}")
        has_any_auth = True
    else:
        _check("🗝️", "XAI_API_KEY not set", False)

    print()

    if not has_any_auth:
        print(_c("red", "  ⚠ No xAI credentials found. Choose one:"))
        print()
        print(f"    {_c('yellow', 'Option A: Browser auth (recommended)')}")
        print(f"      hermes model    → select xAI Grok OAuth")
        print()
        print(f"    {_c('yellow', 'Option B: API key')}")
        print(f"      Get a key from https://console.x.ai/")
        print(f"      Then: hermes config set XAI_API_KEY <your-key>")
        print(f"      Or add XAI_API_KEY=... to {ENV_FILE}")
        print()
        print("  Re-run this script after setting up auth.")
        print()

    # 2. Plugin status
    print(_c("cyan", "▸ Plugins"))
    print()

    plugins = detect_plugin_status()
    for name, info in plugins.items():
        if info["enabled"]:
            detail = "enabled"
        elif info["exists"]:
            detail = "installed (not enabled)"
        else:
            detail = "NOT FOUND"
        _check("🔌", f"{name} (v{info['version']})", info["enabled"], detail)
        if info["exists"] and not info["enabled"]:
            print(f"      → Enable: hermes plugins enable {name}")

    print()

    # 3. Toolset status
    print(_c("cyan", "▸ Toolsets"))
    print()
    print(f"  Run this to check & enable:")
    print(f"    hermes tools list          # see current status")
    print(f"    hermes tools enable image_gen   # enable image gen")
    print(f"    hermes tools enable video_gen   # enable video gen")
    print()

    # 4. End-to-end test (dry run — no actual API call)
    print(_c("cyan", "▸ Readiness Summary"))
    print()

    plugin_ok = plugins.get("image_gen/xai", {}).get("exists", False)

    if has_any_auth and plugin_ok:
        print(_c("green", "  ✅ Looks good! Try generating an image:"))
        print()
        print(f"    hermes -z \"猫の侍を墨絵スタイルで描いて\"")
        print()
        if not oauth:
            print(_c("yellow", "  💡 Tip: You're using an API key. Consider switching to OAuth"))
            print(_c("yellow", "     (browser auth) for a smoother experience: hermes model"))
            print()
    elif has_any_auth and not plugin_ok:
        print(_c("yellow", "  ⚠ Auth is set but the image_gen/xai plugin is not installed."))
        print(_c("yellow", "    Enable it with: hermes plugins enable image_gen/xai"))
        print()
    elif not has_any_auth and plugin_ok:
        print(_c("yellow", "  ⚠ Plugin is installed but no credentials configured."))
        print(_c("yellow", "    See auth options above."))
        print()
    else:
        print(_c("red", "  ✗ Both auth and plugin need setup. See instructions above."))
        print()

    # 5. Also check video_gen
    video_plugin = plugins.get("video_gen/xai", {})
    if video_plugin.get("exists") and not video_plugin.get("enabled"):
        print(f"  {_c('cyan', 'ℹ')} video_gen/xai plugin is also available.")
        print(f"    Enable: hermes plugins enable video_gen/xai")
        print(f"    Then:   hermes tools enable video_gen")
        print()
    elif video_plugin.get("enabled"):
        print(f"  {_c('green', 'ℹ')} video_gen/xai is also enabled — you can generate videos too!")
        print()

    print(_c("bold", "═══════════════════════════════════"))
    print()


if __name__ == "__main__":
    main()
